### ISSUE TITLE
Socket hang up with ExpressJS

### ISSUE THREAD

**Issue by namhq1989** [elastic/apm-agent-nodejs#256]
I've setup my server with `elastic-apm-node`, and it run ok. But after every period of time (30-60 min), it causes an error like below:

```javascript
Error: socket hang up
    at createHangUpError (_http_client.js:331:15)
    at Socket.socketOnEnd (_http_client.js:423:23)
    at emitNone (events.js:111:20)
    at Socket.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1055:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickDomainCallback (internal/process/next_tick.js:218:9)
```

This error made my server down too, so I need to restart it. After removed the code, everthing okay now.
This is my config code:

```javascript
require('elastic-apm-node').start({
  serviceName: '...',
  secretToken: '...',
  instrument: true,
  captureBody: 'all',
  errorOnAbortedRequests: true,
  serverUrl: 'http://...:8200',
  active: process.env.NODE_ENV === 'production'
})
```

Version

```
Elasticsearch, Kibana, APM v6.2.1
NodeJS v8.9.4
ExpressJS v4.15.2
```

I searched for this error, but there's no results with apm-server. Is this a known issue, or my configuration miss something? Please help me to fix this.
Thanks!

**Comment by 7agustibm** [comment]
I had the same problem. I configured APM Server with these values and I solved it.

apm-server.yml:

`apm-server:
  host: 0.0.0.0:8200
  
output.elasticsearch:
  hosts: ['elasticsearch:9200']
`

**Comment by namhq1989** [comment]
Thank you for suggested, but my apm-server.yml config is the same as you mentioned above.

```
apm-server:

  # Defines the host and port the server is listening on
  host: "0.0.0.0:8200"

  # Please be aware that frontend support is an experimental feature at the moment!
  frontend:
    # To enable experimental frontend support set this to true.
    enabled: false

  # Authorization token to be checked. If a token is set here the agents must
  # send their token in the following format: Authorization: Bearer <secret-token>
  secret_token: "..."

  #ssl.enabled: false
  #ssl.certificate : "path/to/cert"
  #ssl.key : "path/to/private_key"

output.elasticsearch:
  # Array of hosts to connect to.
  hosts: ["localhost:9200"]

  # Optional protocol and basic auth credentials.
  #protocol: "https"
  username: "elastic"
  password: "..."
```

This problem only happen in production mode, I tried to simulate random request (0.1-0.5s/request), but my local server running without any errors.
Anyone can help me solve this?

**Comment by watson** [comment]
@namhq1989 Thanks for reporting this.

First of all, the Node agent shouldn't fail with an error like this - even if it does encounter it. It should be resilient to these things. I'll investigate...

But to figure out why you get the socket error in the first place, I need to know a bit more about your production setup. I assume from what you've written that you're running the APM Server and the Node app on two different machines right? How many requests per minute do your production server have?

**Comment by namhq1989** [comment]
@watson Yes I'm running ELK stack on separate machine
Last tracking data is about 117k requests/hour in our main working time (08:00 - 23:00). It may more and more requests at night time.
I turned off apm-server track after 11:00 because of socket hang up error.

<img width="446" alt="screen shot 2018-02-23 at 5 46 57 pm" src="https://user-images.githubusercontent.com/1388374/36590642-e2afba3a-18c1-11e8-9b73-fea8244b9de5.png">

**Comment by watson** [comment]
@namhq1989 thanks for the details. How many processes do you have that are configured with an APM agent?

**Comment by watson** [comment]
@namhq1989 oh, and are they all configured to talk to the same APM Server instance?

**Comment by namhq1989** [comment]
@watson My server is running with only 1 process for my app (system 6GB RAM, 6 cores), and talk to only 1 APM server (system 6GM RAM, 6 cores too). This is 1-1 relationship.

**Comment by watson** [comment]
@namhq1989 hmmm that should be fine. I was working with the theory that the APM Server wasn't able to keep up with the data, but unless it's your Elasticsearch cluster that can't keep up then that doesn't seem to be the issue. Are you seeing any 503 errors in your application logs prior to the error? Also, which version of Node.js are you running?

**Comment by namhq1989** [comment]
@watson There are no 503 or no other errors except `socket hang up`, my server still running but request cannot be process, it did not return anything to client (hanged)

Version
```
Elasticsearch, Kibana, APM v6.2.1
NodeJS v8.9.4
ExpressJS v4.15.2
elastic-apm-node v1.0.3
```

**Comment by watson** [comment]
If there's a socket error when talking to the APM Server, the stack trace will be logged as an error:

https://github.com/elastic/apm-agent-nodejs/blob/b6c3bed3230c205abfea8c8f2d47bfa1bbb76428/lib/request.js#L24-L26

This shouldn't have any impact on your Node process though.

If I understand you correctly, you're saying that the node process keeps running but that it becomes unresponsive?

Does the APM Server by any chance crash or does that keep running as well?

**Comment by namhq1989** [comment]
My server did not crash, but it keeps showing `socket hang up` error, server still receive requests from client, but unresponsive.
```
normal: POST /loginWithPhoneNumber 200 19.889 ms - 77
error: POST /loginWithPhoneNumber 200 -- ms - -
```


I will temporary put `console.log(err)` at line 26 to trace error first.

Btw, I saw any console logs like this:
```
logging error e359ddf9-53ab-4e2b-902e-419b1797377a with Elastic APM
```
but it does not exists in Kibana discover

**Comment by namhq1989** [comment]
@watson  still meet that error

```javascript
 { Error: socket hang up
    at createHangUpError (_http_client.js:331:15)
    at Socket.socketOnEnd (_http_client.js:423:23)
    at emitNone (events.js:111:20)
    at Socket.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1055:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickDomainCallback (internal/process/next_tick.js:218:9) code: 'ECONNRESET' } 
```

should I try ELK 6.1?

**Comment by namhq1989** [comment]
I've simulated with 2 requests/second, and after 20 mins server was crashed

```
<--- Last few GCs --->

[54828:0x103800000]  1045359 ms: Mark-sweep 1278.4 (1431.0) -> 1278.3 (1431.5) MB, 1793.6 / 0.8 ms  allocation failure GC in old space requested
[54828:0x103800000]  1047138 ms: Mark-sweep 1278.3 (1431.5) -> 1278.3 (1424.5) MB, 1779.0 / 1.0 ms  last resort GC in old space requested
[54828:0x103800000]  1049233 ms: Mark-sweep 1278.3 (1424.5) -> 1278.3 (1424.5) MB, 2094.5 / 0.8 ms  last resort GC in old space requested


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x158e8dda5749 <JSObject>
    1: fromString(aka fromString) [buffer.js:~326] [pc=0x22d86935dc28](this=0x158e37d822d1 <undefined>,string=0x158e957fae49 <Very long string[1823926]>,encoding=0x158e8ddb5729 <String[4]: utf8>)
    2: from [buffer.js:193] [bytecode=0x158e4623be11 offset=14](this=0x158e162bba11 <JSFunction Buffer (sfi = 0x158e8ddfeb69)>,value=0x158e957fae49 <Very long string[1823926]>,encodingOrOffset=0x158e8dd...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/usr/local/bin/node]
 2: node::FatalException(v8::Isolate*, v8::Local<v8::Value>, v8::Local<v8::Message>) [/usr/local/bin/node]
 3: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/usr/local/bin/node]
 4: v8::internal::Factory::NewRawTwoByteString(int, v8::internal::PretenureFlag) [/usr/local/bin/node]
 5: v8::internal::String::SlowFlatten(v8::internal::Handle<v8::internal::ConsString>, v8::internal::PretenureFlag) [/usr/local/bin/node]
 6: v8::String::WriteUtf8(char*, int, int*, int) const [/usr/local/bin/node]
 7: node::StringBytes::Write(v8::Isolate*, char*, unsigned long, v8::Local<v8::Value>, node::encoding, int*) [/usr/local/bin/node]
 8: node::Buffer::New(v8::Isolate*, v8::Local<v8::String>, node::encoding) [/usr/local/bin/node]
 9: node::Buffer::(anonymous namespace)::CreateFromString(v8::FunctionCallbackInfo<v8::Value> const&) [/usr/local/bin/node]
10: 0x22d868f380c7
[nodemon] app crashed - waiting for file changes before starting...
```

This is debug logs, there's too many `recovering from wrong currentTransaction`

```terminal
  elastic-apm intercepted mongodb-core.Cursor.prototype._find callback { id: '03bc285c-fa0e-419f-88cd-f99f6110c464' } +0ms
  elastic-apm recovering from wrong currentTransaction { wrong: '86c668b9-1098-430b-8842-bd069b64152d', correct: '03bc285c-fa0e-419f-88cd-f99f6110c464' } +0ms
  elastic-apm ended span { id: '03bc285c-fa0e-419f-88cd-f99f6110c464', name: 'zodyapp-dev.photos.find', type: 'db.mongodb.query', truncated: false } +1ms
  elastic-apm intercepted mongodb-core.Cursor.prototype._find callback { id: '940e4529-ee88-45e2-a583-5007f443aa59' } +1ms
  elastic-apm recovering from wrong currentTransaction { wrong: '86c668b9-1098-430b-8842-bd069b64152d', correct: '940e4529-ee88-45e2-a583-5007f443aa59' } +1ms
  elastic-apm ended span { id: '940e4529-ee88-45e2-a583-5007f443aa59', name: 'zodyapp-dev.photos.find', type: 'db.mongodb.query', truncated: false } +0ms
  elastic-apm intercepted mongodb-core.Cursor.prototype._find callback { id: 'e801b634-253b-45c0-94fc-6ef6b2599742' } +8s
  elastic-apm recovering from wrong currentTransaction { wrong: '86c668b9-1098-430b-8842-bd069b64152d', correct: 'e801b634-253b-45c0-94fc-6ef6b2599742' } +8s
  elastic-apm ended span { id: 'e801b634-253b-45c0-94fc-6ef6b2599742', name: 'zodyapp-dev.configurations.find', type: 'db.mongodb.query', truncated: false } +8s
  elastic-apm intercepted mongodb-core.Cursor.prototype._find callback { id: 'e801b634-253b-45c0-94fc-6ef6b2599742' } +2ms
  elastic-apm recovering from wrong currentTransaction { wrong: '86c668b9-1098-430b-8842-bd069b64152d', correct: 'e801b634-253b-45c0-94fc-6ef6b2599742' } +1ms
  elastic-apm ended span { id: 'e801b634-253b-45c0-94fc-6ef6b2599742', name: 'zodyapp-dev.configurations.find', type: 'db.mongodb.query', truncated: false } +1ms
  elastic-apm flushing queue +14s
  elastic-apm sending transactions payload +18s
```

**Comment by watson** [comment]
I think the memory leak that causes the Node server to crash and the socket timeouts are two different bugs. Do you mind sharing the list of dependencies from your package.json?

**Comment by Sonosar** [comment]
I have the same issue, some details that might help debug.

Used Custom Transactions to track socket.io transactions.
Everything works when custom transactions are not used.

**Comment by namhq1989** [comment]
@Sonosar Thank you but I don't use Custom Transactions for socket.io. I'll try ELK 6.1 instead of latest version.

**Comment by watson** [comment]
After a lot of research into this, we think this is related to an issue with the APM Server that occurs if it's overloaded. In that case the TCP sockets can be kept open because the server hangs on processing the data that the agent is sending to it. Since the agent by default makes a new request to the server every 10 seconds, this might result in a lot of open sockets, which could eat up all the memory. At least that's what we've been able to reproduce.

Because of this we're introducing a new config option `serverTimeout`. This will default to 30 seconds of inactivity, before the socket is terminated: #238

We're also changing the default value of `maxQueueSize` to `100`: #270

We're of course also working on fixing the issue in the APM Server that causes this. And in general working on reducing the resources used by both the APM Server and the agent.

If this is indeed the issue you're seeing, I think your best solution at the moment is to set a low `maxQueueSize` and to deploy more APM Servers behind a load balancer to be able to handle the load.

Let me know if this would be an acceptable solution for you and if it fixes the issue.

**Comment by namhq1989** [comment]
Thank you for helping me, I tried with new version of APM, but still met that error.

```
ELASTIC_APM_SERVICE_NAME = ...
ELASTIC_APM_SECRET_TOKEN = ...
ELASTIC_APM_SERVER_URL = ...
ELASTIC_APM_ACTIVE = true
ELASTIC_APM_CAPTURE_BODY = all
ELASTIC_APM_ERROR_ON_ABORTED_REQUESTS = true
ELASTIC_APM_INSTRUMENT = true
ELASTIC_APM_MAX_QUEUE_SIZE = 50
```

My dev server working perfectly without any errors. Production env have a lot of socketio requests, so I think socket caused error. I'll reorganize my server and track it again.

**Comment by watson** [comment]
@namhq1989 We just released version 1.2.1 of the agent which fixes a memory leak in the agent that could very well be related to the issue you're seeing. Could you try that version?

**Comment by namhq1989** [comment]
@watson Thank you, my package.json agent node version is 1.1.1, I've just upgraded it 4 days ago. I'll try 1.2.1, hope this will solve socket error

**Comment by namhq1989** [comment]
@watson After upgraded agent to 1.2.1, it worked now. But when I do an action to insert single document to database (MongoDB, insert ~10k documents), with interval 100ms each doc, made my server cpu to ~100%, that error `Socket hang up` appear again.
Then I tested 2 times, but it did not appear. So hope you guys can reproduce it.
Thank you for helping me.

**Comment by jppellerin** [comment]
To chime in here: We were seeing the same `Socket hang up` error on version 1.1.1, running a `hapi` server.  After updating to version 1.2.1, we have not seen the error again.  Mind you, we have cut down our sample rate to 0.5.

For reference purposes, here is our config:

```js
import Config from "./config/configuration";

import * as APM from "elastic-apm-node";
APM.start({
    active: Config.get("/elasticSearch/apm/active"),
    serviceName: Config.get("/elasticSearch/apm/serviceName"),
    serverUrl: Config.get("/elasticSearch/apm/url"),
    transactionSampleRate: 0.5
});
```

**Comment by tusharraut1994** [comment]
@watson 

I am also facing socker hang up issue:

Error: socket hang up
    at createHangUpError (_http_client.js:254:15)
    at Socket.socketOnEnd (_http_client.js:346:23)
    at emitNone (events.js:91:20)
    at Socket.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:974:12)
    at elasticAPMCallbackWrapper (/root/ashnikmart/node_modules/elastic-apm-node/lib/instrumentation/index.js:143:27)
    at _combinedTickCallback (internal/process/next_tick.js:80:11)
    at process._tickCallback (internal/process/next_tick.js:104:9)


My Setup details are as follows:
1. Using 3 node elastic cluster.
2. Kibana + APM server + Application are on same instance.

Application is running but in between it gives error as above.

APM settings in nodejs application are as follows:
var apm = require('elastic-apm-node').start({
    serviceName: 'nodejs-application',
    secretToken: 'nodejs',
    serverUrl: 'http://localhost:8200'
})

**Comment by Qard** [comment]
Closing due to inactivity. Please leave a comment if you are still having issues with this.

**Comment by g4br13lg4l4n** [comment]
I had this problem when installed express-fileupload

**Comment by watson** [comment]
@g4br13lg4l4n This issue is related to APM Server v6.2.x and v1.x of the Node.js agent. If you're using this combination I recommend that you upgrade to a newer version, if possible, or use the `serverTimeout` agent config option.

If the issue is unrelated to this, please open a new issue 🙂

**Comment by brdacost** [comment]
I am having more or less the same problem. I am not actually using APM Server nor getting node itself to throw me the Socket Hang Up Error, but I've built this app which listens to a specified port and when the request lasts for more than 2 minutes the client itself gets the Socket Hang Up Error, but my code keeps on running without any errors. However, since the client got the error, even when I finish everything I need to and send him a res.render(), obviously the client doesn't get any response.
I am not entirely sure if this is the appropriated place for asking this, but this was the closest I got to someone with this "same" problem.
